lib/index.js

```diff
@@ -0,0 +1,638 @@
+  processor.Parser = undefined
+  processor.Compiler = undefined
+
+  // Lock.
+  processor.freeze = freeze
+
+  // Plugins.
+  processor.attachers = attachers
+  // @ts-expect-error: overloads are handled.
+  processor.use = use
+
+  // API.
+  processor.parse = parse
+  processor.stringify = stringify
+  // @ts-expect-error: overloads are handled.
+  processor.run = run
+  processor.runSync = runSync
+  // @ts-expect-error: overloads are handled.
+  processor.process = process
+  processor.processSync = processSync
+
+  // Expose.
+  return processor
+
+  // Create a new processor based on the processor in the current scope.
+  /** @type {Processor} */
+  function processor() {
+    const destination = base()
+    let index = -1
+
+    while (++index < attachers.length) {
+      destination.use(...attachers[index])
+    }
+
+    destination.data(extend(true, {}, namespace))
+
+    return destination
+  }
+
+  /**
+   * @param {string|Record<string, unknown>} [key]
+   * @param {unknown} [value]
+   * @returns {unknown}
+   */
+  function data(key, value) {
+    if (typeof key === 'string') {
+      // Set `key`.
+      if (arguments.length === 2) {
+        assertUnfrozen('data', frozen)
+        namespace[key] = value
+        return processor
+      }
+
+      // Get `key`.
+      return (own.call(namespace, key) && namespace[key]) || null
+    }
+
+    // Set space.
+    if (key) {
+      assertUnfrozen('data', frozen)
+      namespace = key
+      return processor
+    }
+
+    // Get space.
+    return namespace
+  }
+
+  /** @type {Processor['freeze']} */
+  function freeze() {
+    if (frozen) {
+      return processor
+    }
+
+    while (++freezeIndex < attachers.length) {
+      const [attacher, ...options] = attachers[freezeIndex]
+
+      if (options[0] === false) {
+        continue
+      }
+
+      if (options[0] === true) {
+        options[1] = undefined
+      }
+
+      /** @type {Transformer|void} */
+      const transformer = attacher.call(processor, ...options)
+
+      if (typeof transformer === 'function') {
+        transformers.use(transformer)
+      }
+    }
+
+    frozen = true
+    freezeIndex = Number.POSITIVE_INFINITY
+
+    return processor
+  }
+
+  /**
+   * @param {Pluggable|null|undefined} [value]
+   * @param {...unknown} options
+   * @returns {Processor}
+   */
+  function use(value, ...options) {
+    /** @type {Record<string, unknown>|undefined} */
+    let settings
+
+    assertUnfrozen('use', frozen)
+
+    if (value === null || value === undefined) {
+      // Empty.
+    } else if (typeof value === 'function') {
+      addPlugin(value, ...options)
+    } else if (typeof value === 'object') {
+      if (Array.isArray(value)) {
+        addList(value)
+      } else {
+        addPreset(value)
+      }
+    } else {
+      throw new TypeError('Expected usable value, not `' + value + '`')
+    }
+
+    if (settings) {
+      namespace.settings = extend(namespace.settings || {}, settings)
```

Comment:
I think it’s better to use a builtin function (`Object.assign()`) instead of a library(`extend`) for shallow merging.

`extend` is used for deep merging in some places, so it can’t be removed as a dependency, but if it ever does become redundant, it’s easier to spot.

```suggestion
      namespace.settings = Object.assign(namespace.settings || {}, settings)
```

Reply:
Agreed